### DISCUSSION TITLE
Is there a way to access the return value of a route callback in middleware, instead of the `starlette.middleware.base._StreamingResponse` converted by FastAPI?

### DISCUSSION THREAD

**Discussion by lovemilk2333** [fastapi/fastapi#12945] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Request, FastAPI
from fastapi.responses import Response

from http import HTTPStatus
from typing import Any, Mapping
from pydantic import BaseModel, Field
from fastapi.responses import JSONResponse


def basemodel2response(status_code: int, model: BaseModel, *, headers=None) -> JSONResponse:
    return JSONResponse(
        status_code=status_code,
        content=model.model_dump(mode='json'),
        headers=headers,
        media_type=JSONResponse.media_type
    )


class BaseResponse(BaseModel):
    code: int
    message: str | None = None
    data: Any | None = None
    headers: Mapping[str, Any] | None = Field(default=None, exclude=True)

    def model_post_init(self, __context: Any) -> None:
        if self.message is None:
            self.message = HTTPStatus(self.code).phrase

    def to_response(self) -> JSONResponse:
        return basemodel2response(self.code, self, headers=self.headers)


def add_response_middleware(
        app: FastAPI, *,
        handle_sequence_and_dict: bool = True,
        handle_sqlmodel: bool = True
):
    from typing import Sequence

    async def _response_middleware(request: Request, call_next):
        response = await call_next(request)
        print(response, isinstance(response, BaseResponse))
        if isinstance(response, Response):
            return response
        elif isinstance(response, BaseResponse):
            return response.to_response()
        elif handle_sequence_and_dict and isinstance(response, (Sequence, dict)):
            return BaseResponse(code=200, data=response).to_response()

        return response

    app.middleware('http')(_response_middleware)


app = FastAPI()
add_response_middleware(app)


@app.get('/test')
async def test():
    return BaseResponse(code=400, message='test')
```

### Description

I want to handle Pydantic BaseModel `BaseResponse` by myself because I like define a class use class variable, but FastAPI auto converted `BaseResponse` to `starlette.middleware.base._StreamingResponse` so that I can't get the original `BaseResponse` object to handle.

### Operating System

Windows

### Operating System Details

版本	Windows 11 专业版
版本号	23H2
安装日期	‎2023/‎10/‎26
操作系统版本	22631.4460
体验	Windows Feature Experience Pack 1000.22700.1047.0

### FastAPI Version

0.115.4

### Pydantic Version

2.9.2

### Python Version

Python 3.10.9

### Additional Context

Sorry my english is bad :(
Can FastAPI provide a option to disable globally the automatic conversion of Pydantic BaseModel parameters?

**Comment by luzzodev** [comment, marked as answer]
Is not possible to access to the original Response subclass in middleware because the response has already been rendered.

Here you have some options to achive the same result ( handle BaseResponse in a custom way)

- Use a custom Response and override render method.
  - Note: if you declare your custom response in method or in router and you don't explicitly return it, serialization will happen before  generating your custom response and you will not solve it.
- Use a custom object (no child of BaseModel, dataclass), write a custom encoder and add it  to `ENCODERS_BY_TYPE`
- Write a simple decorator that you can use to pre-serialize BaseResponse in each endpoint you need.
